server/src/main/java/org/elasticsearch/cluster/ClusterModule.java

```diff
@@ -399,6 +394,7 @@ protected void configure() {
         bind(TaskResultsService.class).asEagerSingleton();
         bind(AllocationDeciders.class).toInstance(allocationDeciders);
         bind(ShardsAllocator.class).toInstance(shardsAllocator);
+        bind(DesiredBalanceShardsAllocator.class).toInstance(desiredBalanceShardsAllocator);
```

Comment:
I don't think we want to create this if it's not configured. Instead we should use the instance provided for `ShardsAllocator.class`, and try to cast it to a `DesiredBalanceShardsAllocator` where necessary.